File: src/14.ts
// Computes the number that produces the longest Collatz chain under a given limit.
//
// See {@link https://projecteuler.net/problem=14}.
const memo = new Map<number, number>();

function collatzLength(n: number): number {
  if (n === 1) {
    return 0;
  }

  if (memo.has(n)) {
    return memo.get(n)!;
  }

  const next = n % 2 === 0 ? n / 2 : 3 * n + 1;
  const result = 1 + collatzLength(next);

  memo.set(n, result);
  return result;
}

export default function compute(limit: number) {
  let max = 0;
  let result = 0;

  for (let i = 1; i < limit; i++) {
    const current = collatzLength(i);
    if (current > max) {
      max = current;
      result = i;
    }
  }

  return result;
}

File: project_euler/14.py
# Problem 14: Longest Collatz Sequence
#
# Which starting number, under one million, produces the longest Collatz sequence?
#
# See https://projecteuler.net/problem=14
from functools import cache


@cache
def collatz_len(n: int) -> int:
    if n == 1:
        return 0

    next = n // 2 if n % 2 == 0 else 3 * n + 1
    return 1 + collatz_len(next)


def run(limit: int) -> int:
    max_len = 0
    current_len = 0
    result = 1

    for i in range(1, limit):
        current_len = collatz_len(i)
        if current_len > max_len:
            max_len = current_len
            result = i

    return result


def test_run():
    assert run(1_000_000) == 837799
